### PR TITLE
Restoring mysql backups from s3

### PR DESCRIPTION
**Pull request by surminus**
This adds the Puppet and tools to enable restoring MySQL backups from an S3 bucket.

Please see commits for detail.

### CODE REVIEW

**Comment by surminus** [comment]
Thank you, rebased and repushed.

File: modules/govuk_mysql/templates/s3cfg.erb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[default]
+bucket_location = eu-west-1
```
Comment (deanwilson) [inline modules/govuk_mysql/templates/s3cfg.erb]: Should this be dynamic? The define allows it to be set to regions other than eu-west-1

File: modules/govuk_mysql/templates/usr/local/bin/xtrabackup_s3_restore [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+# Start the service
+service mysql start
+
+# Clean up
+echo "Cleaning up"
+rm -rf /var/lib/mysql-bak $BASE_DIR $INCR_DIR
+
+# Re-enable Puppet and run it to ensure passwords are updated
+govuk_puppet --enable && govuk_puppet
```
Comment (deanwilson) [inline modules/govuk_mysql/templates/usr/local/bin/xtrabackup_s3_restore]: Should this always re-enable puppet? Even if it wasn't the reason the disable is in place?

**Comment by deanwilson** [comment]
Other than a couple of comments the code looks good.

File: modules/govuk_mysql/templates/usr/local/bin/xtrabackup_s3_restore (hunk 2 above)
Comment (deanwilson) [inline modules/govuk_mysql/templates/usr/local/bin/xtrabackup_s3_restore]: We have a possible solution to this, we're going to look at extending `govuk_puppet` to have a directory of locks so multiple 'things' can disable puppet at the same time and it'll stay disabled until everything has cleared its locks. I'm unsure if this extra work should block this ticket or not.

**Comment by surminus** [comment]
We should wait until [this story](https://trello.com/c/z5wIKSLJ/124-modify-govuk-puppet-to-allow-multiple-locks) is complete and then it can be correctly refactored

File: modules/govuk_mysql/manifests/xtrabackup/restore.pp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  $aws_secret_access_key,
+  $s3_bucket_name,
+  $encryption_key,
+  $aws_region = 'eu-west-1',
+) {
+  $env_sync_envdir = '/etc/mysql/xtrabackup/env_sync/env.d'
+
+  file {[
+        '/etc/mysql/xtrabackup',
+        '/etc/mysql/xtrabackup',
```
Comment (deanwilson) [inline modules/govuk_mysql/manifests/xtrabackup/restore.pp]: Remove the duplicated path.

File: modules/govuk_mysql/manifests/xtrabackup/restore.pp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+  file { "${env_sync_envdir}/AWS_SECRET_ACCESS_KEY":
+    ensure  => 'present',
+    content => $aws_secret_access_key,
+    mode    => '0440',
+  }
+
+  file { '/usr/local/bin/xtrabackup_s3_restore':
+    ensure  => 'present',
+    content => template('govuk_mysql/usr/local/bin/xtrabackup_s3_restore'),
```
Comment (deanwilson) [inline modules/govuk_mysql/manifests/xtrabackup/restore.pp]: This should be an ERB file to indicate it's a template.

File: modules/govuk_mysql/templates/usr/local/bin/xtrabackup_s3_restore [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+DATE=$(date +%s)
+BASE_DIR=/tmp/mysql/basedir/$DATE/
+INCR_DIR=/tmp/mysql/incremental/$DATE/
+
+echo "Starting MySQL environment sync"
+
+# Create directories to restore into
+for i in $BASE_DIR $INCR_DIR; do
+  mkdir -p $i
```
Comment (deanwilson) [inline modules/govuk_mysql/templates/usr/local/bin/xtrabackup_s3_restore]: Can we double quote this please?

File: modules/govuk_mysql/templates/usr/local/bin/xtrabackup_s3_restore [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+done
+
+# Pull the latest backups from S3
+envdir <%= @env_sync_envdir %> gof3r get --endpoint s3-<%= @aws_region -%>.amazonaws.com -b <%= @s3_bucket_name %> -k latest/base.xbcrypt | xbcrypt --encrypt-algo=AES256 --decrypt --encrypt-key="<%= @encryption_key %>" | xbstream -x -C $BASE_DIR
+
+envdir <%= @env_sync_envdir %> gof3r get --endpoint s3-<%= @aws_region -%>.amazonaws.com -b <%= @s3_bucket_name %> -k latest/incremental.xbcrypt | xbcrypt --encrypt-algo=AES256 --decrypt --encrypt-key="<%= @encryption_key %>" | xbstream -x -C $INCR_DIR
+
+# Decompress the backups
+for i in $BASE_DIR $INCR_DIR; do
+  innobackupex --decompress $i
```
Comment (deanwilson) [inline modules/govuk_mysql/templates/usr/local/bin/xtrabackup_s3_restore]: Please double quote this.

File: modules/govuk_mysql/templates/usr/local/bin/xtrabackup_s3_restore [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+innobackupex --apply-log --redo-only $BASE_DIR
+innobackupex --apply-log $BASE_DIR --incremental-dir=$INCR_DIR
+
+# Disable Puppet because it will interfere with the next steps if it runs during
+# this period
+test -f /var/lib/puppet/state/agent_disabled.lock || govuk_puppet --disable
+
+# Stop the service before we copy the data across
+service mysql stop
+
```
Comment (deanwilson) [inline modules/govuk_mysql/templates/usr/local/bin/xtrabackup_s3_restore]: Add a check for status and abort if mysql is still running?

File: modules/govuk_mysql/templates/usr/local/bin/xtrabackup_s3_restore [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+# Recreate the datadir
+mkdir /var/lib/mysql
+
+# Copy the restored data
+innobackupex --copy-back $BASE_DIR
+
+# Update to correct permissions
+echo "Correcting permissions"
+chown -R mysql.mysql /var/lib/mysql
```
Comment (deanwilson) [inline modules/govuk_mysql/templates/usr/local/bin/xtrabackup_s3_restore]: . should be ':'

File: modules/govuk_mysql/manifests/xtrabackup/backup.pp [hunk 9]
```diff
@@ -29,74 +29,68 @@
...
+  $base_service_desc = 'MySQL Xtrabackup daily base push'
 
-    file { '/usr/local/bin/xtrabackup_s3_base':
-      ensure  => 'present',
-      content => template('govuk_mysql/usr/local/bin/xtrabackup_s3_base'),
-      mode    => '0755',
-    }
+  file { '/usr/local/bin/xtrabackup_s3_base':
+    ensure  => 'present',
+    content => template('govuk_mysql/usr/local/bin/xtrabackup_s3_base'),
```
Comment (deanwilson) [inline modules/govuk_mysql/manifests/xtrabackup/backup.pp]: This should be an erb filename.

File: modules/govuk_mysql/manifests/xtrabackup/backup.pp [hunk 10]
```diff
@@ -29,74 +29,68 @@
...
+  $incremental_service_desc = 'MySQL Xtrabackup incremental push'
 
-    file { '/usr/local/bin/xtrabackup_s3_incremental':
-      ensure  => 'present',
-      content => template('govuk_mysql/usr/local/bin/xtrabackup_s3_incremental'),
-      mode    => '0755',
-    }
+  file { '/usr/local/bin/xtrabackup_s3_incremental':
+    ensure  => 'present',
+    content => template('govuk_mysql/usr/local/bin/xtrabackup_s3_incremental'),
```
Comment (deanwilson) [inline modules/govuk_mysql/manifests/xtrabackup/backup.pp]: This should be an ERB.

File: modules/govuk_mysql/manifests/xtrabackup/backup.pp [hunk 11]
```diff
@@ -29,74 +29,68 @@
...
+  }
 
-    file { '/etc/cron.d/xtrabackup_s3':
-      ensure  => 'present',
-      content => template('govuk_mysql/etc/cron.d/xtrabackup_s3'),
-      mode    => '0755',
-    }
+  file { '/etc/cron.d/xtrabackup_s3':
+    ensure  => 'present',
+    content => template('govuk_mysql/etc/cron.d/xtrabackup_s3'),
```
Comment (deanwilson) [inline modules/govuk_mysql/manifests/xtrabackup/backup.pp]: This should be an ERB filename.

File: modules/govuk_mysql/spec/defines/govuk_mysql__xtrabackup_backup_spec.rb [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+        :s3_bucket_name        => 'foo',
+        :aws_region            => 'eu-west-1',
+        :encryption_key        => 'thisisnotarealkey'
+      }}
+
+      it { is_expected.to contain_file('/etc/mysql/xtrabackup').with_ensure('directory') }
+      it { is_expected.to contain_file('/etc/mysql/xtrabackup/env.d').with_ensure('directory') }
+      it { is_expected.to contain_file('/etc/mysql/xtrabackup/env.d/AWS_SECRET_ACCESS_KEY').with_content(/walk-on-the-wild-side/) }
+      it { is_expected.to contain_file('/etc/mysql/xtrabackup/env.d/AWS_ACCESS_KEY_ID').with_content(/lou-reed/) }
+      it { is_expected.to contain_file('/usr/local/bin/xtrabackup_s3_base').with_content(/.*thisisnotarealkey.*/) }
```
Comment (deanwilson) [inline modules/govuk_mysql/spec/defines/govuk_mysql__xtrabackup_backup_spec.rb]: The greedy matchers don't need to be present. `(.*)`

**Comment by surminus** [comment]
This is now ready for final review.

File: modules/govuk_mysql/manifests/xtrabackup/backup.pp [hunk 13]
```diff
@@ -88,9 +93,17 @@
     host_name           => $::fqdn,
   }
 
-  file { '/etc/cron.d/xtrabackup_s3':
-    ensure  => 'present',
-    content => template('govuk_mysql/etc/cron.d/xtrabackup_s3'),
-    mode    => '0755',
+  cron::crondotdee { 'xtrabackup_s3_base':
+    command => 'setlock -N /var/run/mysql_xtrabackup /usr/local/bin/xtrabackup_s3_base',
```
Comment (deanwilson) [inline modules/govuk_mysql/manifests/xtrabackup/backup.pp]: Should the setlocks be absolute paths?

**Comment by deanwilson** [comment]
One small comment. Let's get this in and testing in integration.

Merge away

**Comment by surminus** [comment]
Addressed small comment. Merging!

**Pull request merged**